### ISSUE TITLE
Automatically parse tweets into loadshedding changes

### ISSUE THREAD

**Issue by beyarkay** [beyarkay/eskom-calendar#50]
In general it is impossible to automate loadshedding announcements since eskom has no rhyme nor reason to how they are structured. However it might be able to automate ~50% of them.

### Examples of announcements:
- [#POWERALERT1 Breakdowns and persistent shortages of generation capacity necessitate the continuous 
implementation of Stage 4 loadshedding until 05:00 on Sunday, thereafter Stage 3 until 05:00 on Monday](https://twitter.com/Eskom_SA/status/1570735203718688770)
- [#POWERALERT1 Stage 4 loadshedding will continue to be implemented throughout Thursday and Friday, with a 
possibility of lower stages from Saturday morning](https://twitter.com/Eskom_SA/status/1570033811450372097?s=20&t=H-4DeftNrYTPT00OpyUo5Q)
- [#POWERALERT1 Stage 4 loadshedding will regretfully be implemented from 10:00 this morning until 05:00 on Monday morning](https://twitter.com/Eskom_SA/status/1568494585113976835?s=20&t=H-4DeftNrYTPT00OpyUo5Q)
- [#POWERALERT1 Loadshedding has been increased to Stage 4 until 05:00 on Thursday as three Kendal generation 
units tripped. Thereafter it will revert to Stage 2 until midnight Friday.](https://twitter.com/Eskom_SA/status/1569623551853871106?s=20&t=H-4DeftNrYTPT00OpyUo5Q)
- [#Loadshedding #Stage4 A sudden tripping of three Kendal Power Station units (1 920MW maximum generating capacity) has forced the implementation of Stage 4 loadshedding starting at 10:00 THIS morning until 05:00 on Thursday morning. A full statement will be published shortly](https://twitter.com/Eskom_SA/status/1569579293969195015?s=20&t=H-4DeftNrYTPT00OpyUo5Q)
- [#Stage2 #loadshedding will be implemented from 16:00 until 22:00 this evening Schedules can be downloaded from https://loadshedding.eskom.co.za Municipal customers can obtain schedules from their local municipalities](https://twitter.com/Eskom_SA/status/1567135662469533698?s=20&t=H-4DeftNrYTPT00OpyUo5Q)
- [Load Reduction Notice Date: 05 September 2022 #EskomMpumalanga will implement load reduction tomorrow morning, from 05:00 to 07:00, and in the evening from 17:00 to 19:00. Please refer to the attached statement for areas that will be affected.](https://twitter.com/Eskom_SA/status/1566880118060072962?s=20&t=H-4DeftNrYTPT00OpyUo5Q)
- [#POWERALERT1 Eskom cautions the public that the generation system remains constrained and Stage 2 loadshedding might be implemented at short notice during the evening peaks this week](https://twitter.com/Eskom_SA/status/1561681041168994304?s=20&t=H-4DeftNrYTPT00OpyUo5Q)
- [#POWERALERT1 As previously communicated, Stage 2 loadshedding will be implemented at 16:00 - 24:00 this afternoon and at the same time on Thursday afternoon](https://twitter.com/Eskom_SA/status/1559838433920667648?s=20&t=H-4DeftNrYTPT00OpyUo5Q)

So it might be possible to do something quite simple that would watch for tweets matching a pattern, and then automatically suggest the PR that corresponds to these changes. So a tweet announcing load shedding stage 4 from Aug-18 to Aug-19 would automatically adjust `manually_specified.yaml` to represent those changes.

Note that not all tweets will be automatable, as sometimes the formatting goes out the window.

**Comment by thehta** [comment]
Hey Boyd! I recently came across an interesting use of GPT-3 to extract contextual info from text, I've been playing around with a few of Eskom's tweets, and I think it could be promising.

For example, submitting this tweet (and a lot of additional boilerplate + an example, not shown) into the OpenAI playground,
```
Today is Saturday, 17th of September. Loadshedding Stage 4 will commence at 12:00 and will continue until 05:00 on Sunday as three Kendal generation units tripped. Thereafter it will revert to Stage 2 until midnight Monday.
```

GPT-3 is able to return this string
```
{"stage_4_start_time": "2022-09-17T12:00:00Z", "stage_4_end_time": "2022-09-18T05:00:00Z", "stage_2_start_time": "2022-09-18T05:00:00Z", "stage_2_end_time": "2022-09-19T00:00:00Z"}
```

There might be some tweaking and testing required because of the differing nature of the announcements, but I'd be keen to cut a branch if you think this might be viable

**Comment by ColinBanbury** [comment]
There is an API from Eskom that has the current loadshedding stage. You could use this for inspiration:
https://github.com/swartjean/ha-eskom-loadshedding/blob/master/custom_components/eskom_loadshedding/eskom_interface.py

**Comment by thehta** [comment]
> There is an API from Eskom that has the current loadshedding stage.

Yes, the `/GetStatus` endpoint usually works -- it would be a good way of "ground-truthing" the current Stage, and also of notifying subscribers to a Stage change. I guess this would require some additional logic (and a polling service). @beyarkay, we could spin this out to another ticket? Automatically verify the Eskom Twitter forecast via GetStatus call.

**Comment by beyarkay** [comment]
From [@emmanuelchemla](https://github.com/emmanuelchemla):

> Are stages updated manually?
> 
> How about:
> Scrape the info (shortly after) each new tweet from eskom, from pages like [this](https://www.capetown.gov.za/Family%20and%20home/Residential-utility-services/Residential-electricity-services/Load-shedding-and-outages).
> 
> I would be more than happy to help with this (email me).

I'd prefer to keep conversation in the open so that others can contribute or help out. The main problem has been that eskom's 'official' tweets don't have much consistency, so I wouldn't want to 100% rely on an automated system right off the bat. But definitely interested in getting to 99% automation where a bot can turn a tweet/loadshedding update into a PR that a maintainer just has to approve.

@emmanuelchemla let me know what you're thinking and we can hopefully get something started!

**Comment by beyarkay** [comment]
> @beyarkay, we could spin this out to another ticket? "Automatically verify the Eskom Twitter forecast via GetStatus call."

We could, but at the end of the day I trust what's coming from twitter more than what's coming from the Eskom API because I have my doubts about how technically savvy the people behind loadshedding are, and I think its entirely possible that they forget to tell the poor dev maintaining that endpoint about loadshedding, and then `eskom-calendar` is left with a conflict between the two data sources and no definitive way to resolve them.

@thehta I'm happy for you to open the ticket to setup an automation script to monitor and send alerts when there's a discrepancy so we can see how accurate the getStatus endpoint is, but I'm not sure what we would do if twitter said stage 5 and getStatus said stage 4.

**Comment by thehta** [comment]
> We could, but at the end of the day I trust what's coming from twitter more than what's coming from the Eskom API because I have my doubts about how technically savvy the people behind loadshedding are,

Haha this is interesting, I have the opposite take -- I have a feeling that the API might be linked to Eskom's internal loadshedding system and is closer to ground-truth than Twitter. I get your point nonetheless, it opens up the possibility for a discrepancy. I'll keep this in mind in the new issue. :+1:

**Comment by beyarkay** [comment]
Okay cool. I definitely don't know for sure, so it would be good to get some data to prove/disprove our opinions 👍

**Comment by beyarkay** [comment]
Chipping in to say that I played around with the twitter API and came up with [this little script ](https://gist.github.com/beyarkay/f8ea29956fc21d9321c97fe953111cdf) which downloads the most recent eskom/city of cape town tweets. It might be the start of a more automated system to parse tweets.

**Comment by thehta** [comment]
Nice script. At the time I also looked into some alternatives, but the official API seems to be the best option. It's a shame their tweets aren't easier to parse, and that they don't publish their media desk announcements as PDF. Can you isolate pinned tweets with the API? It seems that #POWERALERT1/#LoadsheddingUpdate (or whatever they decide on the day) is the only other indication of a loadshedding schedule tweet.

**Comment by beyarkay** [comment]
Nice idea about checking the pinned tweet, I'll check it out.

wrt the #POWERALERT1/#LoadsheddingUpdate hashtags, even that isn't guaranteed sometimes.
